subprojects/core/src/main/java/org/gradle/api/internal/project/LifecycleAwareProject.java

```diff
@@ -0,0 +1,1180 @@
+        executeAllprojectsAction();
+        return extensibleDynamicObject.invokeMethod(name, args);
+    }
+
+    @Override
+    @Nullable
+    public Object getProperty(String propertyName) {
+        executeAllprojectsAction();
+        return extensibleDynamicObject.getProperty(propertyName);
+    }
+
+    @Nullable
+    @Override
+    public ProjectInternal getParent() {
+        return delegate.getParent();
+    }
+
+    @Override
+    public String getName() {
+        return delegate.getName();
+    }
+
+    @Override
+    public String getDisplayName() {
+        return delegate.getDisplayName();
+    }
+
+    @Nullable
+    @Override
+    public String getDescription() {
+        executeAllprojectsAction();
+        return delegate.getDescription();
+    }
+
+    @Override
+    public void setDescription(@Nullable String description) {
+        executeAllprojectsAction();
+        delegate.setDescription(description);
+    }
+
+    @Override
+    public Object getGroup() {
+        executeAllprojectsAction();
+        return delegate.getGroup();
+    }
+
+    @Override
+    public void setGroup(Object group) {
+        executeAllprojectsAction();
+        delegate.setGroup(group);
+    }
+
+    @Override
+    public Object getVersion() {
+        executeAllprojectsAction();
+        return delegate.getVersion();
+    }
+
+    @Override
+    public void setVersion(Object version) {
+        executeAllprojectsAction();
+        delegate.setVersion(version);
+    }
+
+    @Override
+    public Object getStatus() {
+        executeAllprojectsAction();
+        return delegate.getStatus();
+    }
+
+    @Override
+    public void setStatus(Object status) {
+        executeAllprojectsAction();
+        delegate.setStatus(status);
+    }
+
+    @Nullable
+    @Override
+    public ProjectInternal getParent(ProjectInternal referrer) {
+        return delegate.getParent(referrer);
+    }
+
+    @Override
+    public ProjectInternal getRootProject() {
+        return delegate.getRootProject();
+    }
+
+    @Override
+    public File getRootDir() {
+        return delegate.getRootDir();
+    }
+
+
+    @Override
+    @SuppressWarnings("deprecation")
+    public File getBuildDir() {
+        executeAllprojectsAction();
+        return delegate.getBuildDir();
+    }
+
+    @SuppressWarnings("deprecation")
+    @Override
+    public void setBuildDir(File path) {
+        executeAllprojectsAction();
+        delegate.setBuildDir(path);
+    }
+
+    @SuppressWarnings("deprecation")
+    @Override
+    public void setBuildDir(Object path) {
+        executeAllprojectsAction();
+        delegate.setBuildDir(path);
+    }
+
+    @Override
+    public File getBuildFile() {
+        return delegate.getBuildFile();
+    }
+
+    @Override
+    public ProjectInternal getRootProject(ProjectInternal referrer) {
+        return delegate.getRootProject(referrer);
+    }
+
+    @Override
+    public Project evaluate() {
+        return delegate.evaluate();
```

Comment:
❌  Isn't this mutable access?

The (CrossProjectAccess)`ProblemReportingProject` has `shouldNotBeUsed()` in a number of methods, including this one. It means that we ensured that these methods are not called when IP is enabled. But I am not so certain they cannot be called in vintage, where the lifecycle callbacks should operate normally.

There seem to be a number of other methods like this. To avoid such errors, we should do what @bamboo suggested originally and extract an abstract `MutableAccessTrackingProject` class which both `LifecycleAwareProject` and `ProblemReportingProject` should extend.